### ISSUE TITLE
Need option to parse command-line without changing argc, argv

### ISSUE THREAD

**Issue by eyalroz** [jarro2783/cxxopts#243]
Currently, `options.parse()` modifies argc and argv, and takes non-const references:
```
Options::parse(int& argc, char**& argv)
```
This may often make sense, but sometimes - it doesn't. Parsing may be speculative; or there needs to be something else done with the command-line in addition to parsing, etc. This change is also obviously not necessary - parsing can be done without it. This is particularly inconvenient when you want to pass your argc or argv as non-lvalue-references).

So, I suggest one of the following be done:

1. The default behavior be changed to take an `int` and a `const char**` (or a `char **` if you must), ,with no changes.
2. An additional method, e.g. `parse_and_conserve` or `non_destructive_parse` or `non_consumptive_parse` be introduced.

**Comment by jarro2783** [comment]
This has been brought up a few times before. I have started work on refactoring a few things and this will be part of the changes. Hopefully it will be done in the next couple of months.
See #91, #123, #224.

**Comment by eyalroz** [comment]
@jarro2783 : But isn't this a 1-minute fix? At least if we consider option 2?

```
Options::non_cosumptive_parse(int argc, char** argv) { return parse(argc, argv); }
```

**Comment by jarro2783** [comment]
That wouldn't quite work, since it's not `argv` itself that is changed, but the pointers in the argv array that are actually moved around.
Even if it did something like that, the behaviour could be a bit unintuitive, since you would lose any information about unparsed arguments.
The short fix would be something like this:
```
ParseResult Options::non_cosumptive_parse(int argc, char** argv) {
  std::unique_ptr<char*[]> new_argv = std::make_unique<char*[]>(argc);
  std::copy(argv, argv + argc, new_argv.get());
  return parse(argc, new_argv);
}
```

**Comment by eyalroz** [comment]
Maybe just:
```
ParseResult Options::non_cosumptive_parse(int argc, char** argv) {
  std::vector<char*> new_argv {argv, argv + argc};
  return parse(argc, new_argv);
}

**Comment by vonhezhou** [comment]
+1 for adding an option or another function for this. Consider this:
1. you are using some library that parses argc & argv, e.g.  cef, nodejs
2. your application needs some more option, e.g. config file path, work directory, etc.
then changing argv IS intended.

**Comment by jarro2783** [comment]
This was fixed in the 3.0 release.

**Comment by eyalroz** [comment]
@jarro2783 : Please consider creating "proper" releases on this mini-site's [Releases page](https://github.com/jarro2783/cxxopts/releases), and perhaps even copying the release part of the [changelog](https://github.com/jarro2783/cxxopts/blob/master/CHANGELOG.md) to the release page.
